tests/testmarker/walker.go

```diff
@@ -0,0 +1,373 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"bytes"
+	_ "embed"
+	"fmt"
+	"go/ast"
+	"go/parser"
+	"go/token"
+	"io/fs"
+	"log"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"regexp"
+	"sort"
+	"strconv"
+
+	"github.com/ghemawat/stream"
+	"github.com/zabawaba99/go-gitignore"
+	"gopkg.in/yaml.v2"
+)
+
+var (
+	//go:embed rules.yaml
+	rulesYaml string
+	rules     []*RuleSpec
+
+	// records all the test cases that should be marked
+	newTestCases  map[string][]string = make(map[string][]string)
+	testNameRegex                     = regexp.MustCompile(`^\+func\s+(Test.*)\(`)
+)
+
+// MarkInfo describes the data structure of the feature mapping of a test case
+type MarkInfo struct {
+	Features []featureMarkInfo `yaml:"features,omitempty"`
+	Issues   []issueMarkInfo   `yaml:"issues,omitempty"`
+	// TestName is the name of the test method
+	TestName string `yaml:"testName,omitempty"`
+	// File records the file where the test case is defined
+	File string `yaml:"file"`
+}
+
+type featureMarkInfo struct {
+	ID          string `yaml:"id"`
+	Description []any  `yaml:"description"`
+}
+
+type issueMarkInfo struct {
+	ID       int    `yaml:"id"`
+	IssueURL string `yaml:"url"`
+}
+
+// RuleSpec describes the data structure of the mapping spec
+type RuleSpec struct {
+	Path    string   `yaml:"path"`
+	Exclude []string `yaml:"exclude"`
+}
+
+func main() {
+	fset := token.NewFileSet()
```

Comment:
Please use nogo to implement it. nogo is at ```build/linter```.

Reply:
Using nogo doesn't work for my needs for the following reasons:
1) The QE team needs to cover as many incremental test cases as possible. For the scenario of adding test cases to the existing _test.go file, the nogo_config.json configuration cannot satisfy.
2) The test maker needs to scan git diff to solve the scenario of test case reconstruction: for example, move the test case from one _test.go to another _test.go file. We should not regard the test cases that move in this scenario as new test cases. The way nogo works makes it difficult to solve this problem.

@hawkingrei